### PR TITLE
Networks

### PR DESCRIPTION
**Pull request by yallup**
# Description

Adds basic neural network fitting functionality

# Checklist:

- [x] I have performed a self-review of my own code
- [x] My code is PEP8 compliant (`flake8 lsbi tests`)
- [x] My code contains compliant docstrings (`pydocstyle --convention=numpy lsbi`)
- [x] New and existing unit tests pass locally with my changes (`python -m pytest`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have appropriately incremented the [semantic version number](https://semver.org/) in both README.rst and lsbi/_version.py

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -2,8 +2,8 @@
 lsbi: Linear Simulation Based Inference
 =======================================
 :lsbi: Linear Simulation Based Inference
-:Author: Will Handley
-:Version: 0.2.0
+:Author: Will Handley, David Yallup
+:Version: 0.2.1
```
Comment (williamjameshandley) [inline README.rst]: Definitely 0.3.0!

File: lsbi/network.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        #     counter = 0
+        # else:
+        #     counter += 1
+        #     if counter >= patience:
+        #         print(
+        #             f"No improvement for {patience} epochs. \
+        #                 Rolling back to best epoch."
+        #         )
+        #         model.load_state_dict(best_weights)
+        #         break
```
Comment (williamjameshandley) [inline lsbi/network.py]: Do we need this?

File: lsbi/network.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+def train(
+    model,
+    data,
+    labels,
+    num_epochs=10,
+    batch_size=128,
+    decay_rate=0.95,
+    lr=0.001,
+):
```
Comment (williamjameshandley) [inline lsbi/network.py]: I know a separate 'train' function is how it is done in the deep learning/torch literature, but at this point we could consider making this a scikitlearn-like train/predict member function. At some point (not this PR), we could even wrap in sklearn architecture.

**Review by williamjameshandley — CHANGES_REQUESTED** [review]
Excellent -- some minor suggestions/comments inline.

File: lsbi/network.py (hunk 3 above)
Comment (yallup) [reply lsbi/network.py]: agreed, especially as the loss etc. is folded into the classifier it makes more sense in future to just reject the functional ml workflow in favour of oo

File: lsbi/network.py (hunk 2 above)
Comment (yallup) [reply lsbi/network.py]: no, just uncommented to pass coverage but also probably not needed and can be revisited in the future if we do refactor train

File: lsbi/network.py (hunk 3 above)
Comment (williamjameshandley) [reply lsbi/network.py]: Are we going to do this in this PR?

Comment (yallup) [reply lsbi/network.py]: moved train to the model, now called as `fit(X,y)` a la sklearn

**Review by williamjameshandley — COMMENTED** [review]
Looking much better.

I have a couple of adjustments which I'll now push, so that you can review the diff.

File: lsbi/network.py [hunk 4]
```diff
@@ -6,47 +6,57 @@
...
-        self.dense_2 = nn.Linear(internal_dim, internal_dim)
-        self.leaky_relu_2 = nn.LeakyReLU()
-        self.batch_norm_2 = nn.BatchNorm1d(internal_dim)
-        self.dense_3 = nn.Linear(internal_dim, internal_dim)
-        self.leaky_relu_3 = nn.LeakyReLU()
-        self.batch_norm_3 = nn.BatchNorm1d(internal_dim)
-        self.dense_4 = nn.Linear(internal_dim, internal_dim)
-        self.leaky_relu_4 = nn.LeakyReLU()
-        self.dense_5 = nn.Linear(internal_dim, 1)
+        self.model = nn.Sequential(
```
Comment (williamjameshandley) [inline lsbi/network.py:34]: This felt more 'torchy', and less error prone. It does remove the named batch_norm attributes, which I couldn't quite work out why they were needed.

File: lsbi/network.py [hunk 5]
```diff
@@ -112,13 +132,6 @@ def fit(self, X, y, **kwargs):
             mean_loss = torch.mean(torch.tensor(epoch_loss)).item()
             print(f"Epoch {epoch+1}/{num_epochs}, Loss: {mean_loss}")
 
-        # Evaluate the model
```
Comment (williamjameshandley) [inline lsbi/network.py]: What are these lines doing (no tests break, but they look like they could be important)

**Review by williamjameshandley — APPROVED** [review]

File: README.rst [hunk 6]
```diff
@@ -3,7 +3,7 @@ lsbi: Linear Simulation Based Inference
 =======================================
 :lsbi: Linear Simulation Based Inference
 :Author: Will Handley
```
Comment (williamjameshandley) [inline README.rst]: Add yourself here

File: pyproject.toml [hunk 7]
```diff
@@ -16,6 +17,7 @@ dependencies = [
     'numpy',
     'scipy',
     'matplotlib',
+    'torch',
```
Comment (williamjameshandley) [inline pyproject.toml:20]: At some point we might want to make this optional, for users that only use the linear stuff.

File: lsbi/network.py (hunk 4 above)
Comment (yallup) [reply lsbi/network.py:34]: Yes much nicer! The batch norm attributes don't need to be individually exposed anyway

File: lsbi/network.py (hunk 5 above)
Comment (yallup) [reply lsbi/network.py]: So I thought these were here for a reason and upon checking I think without switching to eval(), batch norm does the wrong thing at inference time, I've added the switch back, but the whole model can just be turned now it is wrapped.

I thought about adding a test for this but the result is not an error just misleading numerical values so it seems a bit overkill to test

File: pyproject.toml (hunk 7 above)
Comment (yallup) [reply pyproject.toml:20]: We can do this now but I didn't have the code to gracefully warn and exit if networks is imported without the optional dependency to hand. If this is easy and I'm just unaware then go ahead with this

Comment (williamjameshandley) [reply pyproject.toml:20]: Let's do it at a later date

**Review by williamjameshandley — APPROVED** [review]
Many thanks @yallup. Please squash and merge.
